### ISSUE TITLE
proptest tries to write to /home/proptest-regressions

### ISSUE THREAD

**Issue by jonas-schievink** [AltSysrq/proptest#86]
I have a very basic setup that doesn't change any proptest settings, and upon running `cargo test` I found this in the output:

```
proptest: failed to append to /home/proptest-regressions/dev/x87/tests/f80.txt: Permission denied (os error 13)
```

Now this happens because there's a `main.rs` in my home directory that I use to quickly test things that the playpen can't. Maybe proptest should impose a limit on where it searches for files, or try to locate the project's `Cargo.toml` instead of `.rs` files?

**Comment by jonas-schievink** [comment]
When I remove `~/main.rs`, running `cargo test` in the project's directory results in proptest complaining that it doesn't know where to put the persistence files:

    proptest: FileFailurePersistence::SourceParallel set, but failed to find lib.rs or main.rs

(other than that, proptest has worked amazingly well - great job!)

**Comment by AltSysrq** [comment]
I recall there is a reason I specifically did not have it look for `Cargo.toml`, but apparently I did not write it down anywhere so I'm not sure what that reason was.

I agree though that the current system has problems. I'm going to need some time to think about what the best option is here.

**Comment by Eh2406** [comment]
Maybe we can ask `cargo` for this information using `cargo metadata`?

**Comment by AltSysrq** [comment]
I don't feel that it's safe to rely on `cargo` being in `$PATH`. I think it would also make it sensitive to the CWD in workspace projects.

**Comment by jonas-schievink** [comment]
Cargo also sets the `$CARGO` environment variable to the Cargo binary performing the build, so maybe you can use that?

**Comment by AltSysrq** [comment]
Yeah, that would take care of the first point, but there's still the issue that it could produce different results from `cargo test` vs `workspace-member/target/debug/workspace-member-deadbeef` (also, if someone runs the test binary directly `$CARGO` might not be set).  Maybe the CWD issue could be handled by running it with the parent of the source file as the CWD though.

**Comment by jonas-schievink** [comment]
The JSON output of `cargo metadata` contains a `workspace_root` field that probably always points to the dir containing the root manifest.

> also, if someone runs the test binary directly $CARGO might not be set

That's true, it a fallback could be to just invoke `cargo` directly.

Note that invoking Cargo basically just tries to locate the nearest `Cargo.toml`, so it will probably run into the same problems you had, so at this point retrying the "look for `Cargo.toml`" approach might be worth a try?

**Comment by AltSysrq** [comment]
> at this point retrying the "look for `Cargo.toml`" approach might be worth a try?

Yeah. I'm starting to think the rationale for not using `Cargo.toml` may have been rooted in an earlier placement strategy. If I can't think of a reason this wouldn't work, I'll probably just switch to the following:

1. Use the current method to derive a path. If `proptest-regressions` exists there, continue using it (for backward compatibility).

2. Search upward from the source file for a `Cargo.toml`. Bail if it notices it's leaving a popular SCM (e.g. don't go to the parent directory if the current contains `.git`). Place `proptest-regressions` in the same directory as `Cargo.toml`.

3. If there's still no path, give up and switch to the `WithSource` configuration.
